File: dev/statichttp.pl
#!/usr/bin/perl
# Мъничък статичен сървър за преглед (само за разработка).
# Пускане:  perl dev/statichttp.pl 8791
use strict; use warnings;
use IO::Socket::INET;
use Cwd qw(abs_path getcwd);

my $port = $ARGV[0] || 8791;
my $root = getcwd();
my %MIME = (
  html => 'text/html; charset=utf-8', js => 'text/javascript; charset=utf-8',
  css  => 'text/css; charset=utf-8',  json => 'application/json; charset=utf-8',
  webmanifest => 'application/manifest+json; charset=utf-8',
  svg => 'image/svg+xml', png => 'image/png', jpg => 'image/jpeg',
  woff2 => 'font/woff2', ico => 'image/x-icon', md => 'text/plain; charset=utf-8',
);
my $srv = IO::Socket::INET->new(LocalAddr => '127.0.0.1', LocalPort => $port,
  Listen => 32, Reuse => 1) or die "не мога да отворя порт $port: $!";
print "static: http://127.0.0.1:$port/ от $root\n";
# 🔴 26.08 — СЪРВЪРЪТ ВИСЕШЕ ЗАВИНАГИ И УБИВАШЕ САМ СЕБЕ СИ.
# Той е ЕДНОНИШКОВ: обслужва по една връзка. А `<$cl>` чака ред и БЛОКИРА,
# докато не дойде. Отвори ли браузърът връзка предварително (preconnect,
# keep-alive) и не прати нищо, сървърът застива на нея — и НИТО ЕДНА следваща
# заявка не се обслужва. Отвън изглежда като „сървърът умря".
# Днес ме спъна ДВА ПЪТИ: веднъж спря да отговаря насред работа, и веднъж
# направи невъзможна проверката за офлайн — регистрацията на обслужващия
# работник тегли sw.js по СВОЯ връзка и увисна.
# ЛЕКЪТ: срок за четене. Не дойде ли заявката за 5 секунди, връзката се
# затваря и сървърът продължава да живее.
use IO::Select;
while (my $cl = $srv->accept) {
  my $sel = IO::Select->new($cl);
  my $req = $sel->can_read(5) ? <$cl> : undef;
  while ($sel->can_read(5) and defined(my $h = <$cl>)) { last if $h =~ /^\r?\n$/; }
  unless (defined $req and $req =~ m{^GET\s+(\S+)}) { close $cl; next; }
  my $path = $1; $path =~ s/\?.*$//; $path =~ s/%([0-9A-Fa-f]{2})/chr(hex($1))/ge;
  $path = '/index.html' if $path eq '/';
  my $file = $root . $path;
  my $ok = 0; my $real = eval { abs_path($file) };
  $ok = 1 if defined $real and index($real, abs_path($root)) == 0 and -f $real;
  if (!$ok) { print $cl "HTTP/1.1 404 Not Found\r\nContent-Length: 0\r\n\r\n"; close $cl; next; }
  my ($ext) = $real =~ /\.([A-Za-z0-9]+)$/; $ext = lc($ext || '');
  my $type = $MIME{$ext} || 'application/octet-stream';
  open(my $fh, '<:raw', $real) or do { close $cl; next; };
  local $/; my $body = <$fh>; close $fh;
  print $cl "HTTP/1.1 200 OK\r\nContent-Type: $type\r\nContent-Length: " . length($body)
          . "\r\nCache-Control: no-store\r\nConnection: close\r\n\r\n";
  print $cl $body;
  close $cl;
}

File: dev/statichttp.js
// ═══════════════════════════════════════════════════════════
// 🌐 СТАТИЧЕН СЪРВЪР ЗА ПРЕГЛЕД — на Node, защото перлът виси
//
// ЗАЩО СЪЩЕСТВУВА (26.08.2026): dev/statichttp.pl е ЕДНОНИШКОВ и днес ме
// спъна ТРИ пъти. Причината не е една, а две, и втората не се лекува с
// кръпка:
//   1. `<$cl>` блокира, докато не дойде ред. Отвори ли браузърът връзка
//      предварително (preconnect, keep-alive) и не прати нищо, сървърът
//      застива на нея и НИТО ЕДНА следваща заявка не се обслужва. Сложих
//      срок за четене (IO::Select) — това го оправи.
//   2. Но `print $cl $body` блокира СЪЩО, когато клиентът не чете достатъчно
//      бързо. Страницата тегли 114 файла; докато машината е натоварена,
//      един бавен клиент спира целия сървър. Отвън изглежда като „умрял".
// Кръпка по кръпка на еднонишков сървър е губене на време. Node обслужва
// връзките едновременно по устройство — целият клас проблем изчезва.
//
// ПУСКАНЕ: node dev/statichttp.js [порт]     (по подразбиране 8791)
// ПЪТ НАЗАД: dev/statichttp.pl стои непокътнат и още работи за прости случаи.
//
// 🪤 КАПАНИТЕ, ОБМИСЛЕНИ:
//   · Излизане от папката: `..` в адреса не бива да дава достъп нагоре.
//     Пътят се нормализира и се проверява, че остава в корена.
//   · Кирилица в имената: адресът се декодира с decodeURIComponent.
//   · Обслужващият работник иска ТОЧЕН тип за sw.js — иначе браузърът
//     отказва да го регистрира.
//   · Кеш: `no-store`, за да не се налага да се гадае дали виждаш новото.
//   · Само GET и HEAD. Нищо не се записва — това е сървър за ГЛЕДАНЕ.
// ═══════════════════════════════════════════════════════════
const http = require('http');
const fs = require('fs');
const path = require('path');

const ПОРТ = Number(process.argv[2]) || 8791;
const КОРЕН = path.resolve(__dirname, '..');

const ТИПОВЕ = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.webmanifest': 'application/manifest+json; charset=utf-8',
  '.svg': 'image/svg+xml',
  '.png': 'image/png', '.jpg': 'image/jpeg', '.jpeg': 'image/jpeg',
  '.gif': 'image/gif', '.webp': 'image/webp',
  '.woff2': 'font/woff2', '.woff': 'font/woff',
  '.ico': 'image/x-icon', '.md': 'text/plain; charset=utf-8',
  '.txt': 'text/plain; charset=utf-8', '.mp3': 'audio/mpeg', '.mp4': 'video/mp4'
};

const сървър = http.createServer((зая, отг) => {
  if (зая.method !== 'GET' && зая.method !== 'HEAD') {
    отг.writeHead(405, { 'Content-Length': 0 }); отг.end(); return;
  }
  // 🪤 `url.parse()` е остарял (Node сам предупреждава: DEP0169, без CVE-та за
  //   дупките му). Новият URL иска пълен адрес, затова се дава основа.
  let път;
  try { път = decodeURIComponent(new URL(зая.url, 'http://x').pathname || '/'); }
  catch (e) { отг.writeHead(400, { 'Content-Length': 0 }); отг.end(); return; }
  if (път === '/' || път === '') път = '/index.html';

  // 🪤 без излизане нагоре: нормализираме и настояваме да сме в корена
  const цел = path.resolve(КОРЕН, '.' + път);
  if (цел !== КОРЕН && !цел.startsWith(КОРЕН + path.sep)) {
    отг.writeHead(403, { 'Content-Length': 0 }); отг.end(); return;
  }

  fs.stat(цел, (гр, ст) => {
    if (гр || !ст.isFile()) { отг.writeHead(404, { 'Content-Length': 0 }); отг.end(); return; }
    const тип = ТИПОВЕ[path.extname(цел).toLowerCase()] || 'application/octet-stream';
    отг.writeHead(200, {
      'Content-Type': тип,
      'Content-Length': ст.size,
      'Cache-Control': 'no-store',
      // обслужващият работник има право да управлява целия сайт
      'Service-Worker-Allowed': '/'
    });
    if (зая.method === 'HEAD') { отг.end(); return; }
    const поток = fs.createReadStream(цел);
    поток.on('error', () => { try { отг.destroy(); } catch (e) {} });
    поток.pipe(отг);
  });
});

// връзка, която не праща нищо, не бива да държи ресурс вечно
сървър.headersTimeout = 10000;
сървър.requestTimeout = 30000;
сървър.keepAliveTimeout = 5000;

сървър.on('error', гр => {
  if (гр.code === 'EADDRINUSE') {
    console.log('🔴 порт ' + ПОРТ + ' е зает. Спри другия сървър или дай друг порт:');
    console.log('   node dev/statichttp.js 8792');
  } else console.log('🔴 ' + гр.message);
  process.exit(1);
});

сървър.listen(ПОРТ, '127.0.0.1', () => {
  console.log('🌐 http://127.0.0.1:' + ПОРТ + '/   от ' + КОРЕН);
  console.log('   (Node — обслужва връзките едновременно; перлената версия висеше)');
});
